<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.6"/>
<title>mapgd: population_genotypes Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">mapgd
   &#160;<span id="projectnumber">0.4</span>
   </div>
   <div id="projectbrief">A program for the Maximum-likelihood analysis of population genomic data.</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.6 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="modules.html"><span>Modules</span></a></li>
      <li class="current"><a href="annotated.html"><span>Data&#160;Structures</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Data&#160;Structures</span></a></li>
      <li><a href="classes.html"><span>Data&#160;Structure&#160;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Data&#160;Fields</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Data Structures</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Friends</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Groups</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Data Fields</a> &#124;
<a href="#pub-static-attribs">Static Public Attributes</a> &#124;
<a href="#pri-attribs">Private Attributes</a>  </div>
  <div class="headertitle">
<div class="title">population_genotypes Class Reference</div>  </div>
</div><!--header-->
<div class="contents">
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a1a01615f1d191f7dcfed5548aca4e6e9"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1a01615f1d191f7dcfed5548aca4e6e9"></a>
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#a1a01615f1d191f7dcfed5548aca4e6e9">population_genotypes</a> ()</td></tr>
<tr class="memdesc:a1a01615f1d191f7dcfed5548aca4e6e9"><td class="mdescLeft">&#160;</td><td class="mdescRight">simple constructor. <br/></td></tr>
<tr class="separator:a1a01615f1d191f7dcfed5548aca4e6e9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad2a84f4704015d323678fd6575f55f79"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad2a84f4704015d323678fd6575f55f79"></a>
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#ad2a84f4704015d323678fd6575f55f79">population_genotypes</a> (std::vector&lt; std::string &gt;)</td></tr>
<tr class="memdesc:ad2a84f4704015d323678fd6575f55f79"><td class="mdescLeft">&#160;</td><td class="mdescRight">constructor needed by map-file. String should be coloumn names. <br/></td></tr>
<tr class="separator:ad2a84f4704015d323678fd6575f55f79"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af447d47e419209584f2edb7d3c7d79a5"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af447d47e419209584f2edb7d3c7d79a5"></a>
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#af447d47e419209584f2edb7d3c7d79a5">population_genotypes</a> (const <a class="el" href="classpopulation__genotypes.html">population_genotypes</a> &amp;)</td></tr>
<tr class="memdesc:af447d47e419209584f2edb7d3c7d79a5"><td class="mdescLeft">&#160;</td><td class="mdescRight">constructor using a population_genotype <br/></td></tr>
<tr class="separator:af447d47e419209584f2edb7d3c7d79a5"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8575634882d02791d9ea7621abe62d7f"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a8575634882d02791d9ea7621abe62d7f"></a>
&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#a8575634882d02791d9ea7621abe62d7f">~population_genotypes</a> ()</td></tr>
<tr class="memdesc:a8575634882d02791d9ea7621abe62d7f"><td class="mdescLeft">&#160;</td><td class="mdescRight">destructor. <br/></td></tr>
<tr class="separator:a8575634882d02791d9ea7621abe62d7f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9fe492e927191dbf60f13e4965494a85"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#a9fe492e927191dbf60f13e4965494a85">size</a> () const </td></tr>
<tr class="memdesc:a9fe492e927191dbf60f13e4965494a85"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns the number of samples.  <a href="#a9fe492e927191dbf60f13e4965494a85">More...</a><br/></td></tr>
<tr class="separator:a9fe492e927191dbf60f13e4965494a85"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a60e44f0b435c3e2611d27b8c64925933"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a60e44f0b435c3e2611d27b8c64925933"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#a60e44f0b435c3e2611d27b8c64925933">add</a> (const <a class="el" href="classgenotype.html">genotype</a> &amp;likelihood)</td></tr>
<tr class="memdesc:a60e44f0b435c3e2611d27b8c64925933"><td class="mdescLeft">&#160;</td><td class="mdescRight">append a sample to the likelihood. <br/></td></tr>
<tr class="separator:a60e44f0b435c3e2611d27b8c64925933"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac09d2532e86a5877a5b6e3f126afdc72"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ac09d2532e86a5877a5b6e3f126afdc72"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#ac09d2532e86a5877a5b6e3f126afdc72">add</a> (const float_t &amp;lMM, const float_t &amp;lMm, const float_t &amp;lmm, const count_t &amp;N)</td></tr>
<tr class="memdesc:ac09d2532e86a5877a5b6e3f126afdc72"><td class="mdescLeft">&#160;</td><td class="mdescRight">append a sample to the likelihood. <br/></td></tr>
<tr class="separator:ac09d2532e86a5877a5b6e3f126afdc72"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9893d7e9f8d08bfab28f26ffeb66d539"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a9893d7e9f8d08bfab28f26ffeb66d539"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#a9893d7e9f8d08bfab28f26ffeb66d539">clear</a> ()</td></tr>
<tr class="memdesc:a9893d7e9f8d08bfab28f26ffeb66d539"><td class="mdescLeft">&#160;</td><td class="mdescRight">clear likelihoods. <br/></td></tr>
<tr class="separator:a9893d7e9f8d08bfab28f26ffeb66d539"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1209d8766d17bf577d390ca25890f8f8"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1209d8766d17bf577d390ca25890f8f8"></a>
std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#a1209d8766d17bf577d390ca25890f8f8">header</a> (void) const </td></tr>
<tr class="memdesc:a1209d8766d17bf577d390ca25890f8f8"><td class="mdescLeft">&#160;</td><td class="mdescRight">print header. <br/></td></tr>
<tr class="separator:a1209d8766d17bf577d390ca25890f8f8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a59dcc8df94153d4a5fac4e13290815c1"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a59dcc8df94153d4a5fac4e13290815c1"></a>
std::vector&lt; std::string &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>get_sample_names</b> (void) const </td></tr>
<tr class="separator:a59dcc8df94153d4a5fac4e13290815c1"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a06bcdc48affdaba3dc415dbb4be99152"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a06bcdc48affdaba3dc415dbb4be99152"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#a06bcdc48affdaba3dc415dbb4be99152">set_sample_names</a> (const std::vector&lt; std::string &gt; &amp;<a class="el" href="classpopulation__genotypes.html#a7a6f79579434313fd1ece9837ff7aa85">sample_names</a>)</td></tr>
<tr class="memdesc:a06bcdc48affdaba3dc415dbb4be99152"><td class="mdescLeft">&#160;</td><td class="mdescRight">names of the samples sequenced. <br/></td></tr>
<tr class="separator:a06bcdc48affdaba3dc415dbb4be99152"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Data Fields</h2></td></tr>
<tr class="memitem:ad336d3284899bc6f3135d951018583aa"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad336d3284899bc6f3135d951018583aa"></a>
std::vector&lt; <a class="el" href="classgenotype.html">genotype</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#ad336d3284899bc6f3135d951018583aa">likelihoods</a></td></tr>
<tr class="memdesc:ad336d3284899bc6f3135d951018583aa"><td class="mdescLeft">&#160;</td><td class="mdescRight">genotypic likelihood <br/></td></tr>
<tr class="separator:ad336d3284899bc6f3135d951018583aa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7a6f79579434313fd1ece9837ff7aa85"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a7a6f79579434313fd1ece9837ff7aa85"></a>
std::vector&lt; std::string &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#a7a6f79579434313fd1ece9837ff7aa85">sample_names</a></td></tr>
<tr class="memdesc:a7a6f79579434313fd1ece9837ff7aa85"><td class="mdescLeft">&#160;</td><td class="mdescRight">genotypic likelihood <br/></td></tr>
<tr class="separator:a7a6f79579434313fd1ece9837ff7aa85"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae06a467d7131072986c1949a45e48dc4"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae06a467d7131072986c1949a45e48dc4"></a>
gt_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#ae06a467d7131072986c1949a45e48dc4">major</a></td></tr>
<tr class="memdesc:ae06a467d7131072986c1949a45e48dc4"><td class="mdescLeft">&#160;</td><td class="mdescRight">identity of the major allele <br/></td></tr>
<tr class="separator:ae06a467d7131072986c1949a45e48dc4"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afb4ccf7c2dba4ad6c13cd736a35b5684"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="afb4ccf7c2dba4ad6c13cd736a35b5684"></a>
gt_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#afb4ccf7c2dba4ad6c13cd736a35b5684">minor</a></td></tr>
<tr class="memdesc:afb4ccf7c2dba4ad6c13cd736a35b5684"><td class="mdescLeft">&#160;</td><td class="mdescRight">identity of the minor allele <br/></td></tr>
<tr class="separator:afb4ccf7c2dba4ad6c13cd736a35b5684"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adec803d2027fdeeb67c8384cd9621364"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="adec803d2027fdeeb67c8384cd9621364"></a>
float_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#adec803d2027fdeeb67c8384cd9621364">m</a></td></tr>
<tr class="memdesc:adec803d2027fdeeb67c8384cd9621364"><td class="mdescLeft">&#160;</td><td class="mdescRight">minor allele frequency <br/></td></tr>
<tr class="separator:adec803d2027fdeeb67c8384cd9621364"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a1c53ffb4e7e847387154745fbb80b74f"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a1c53ffb4e7e847387154745fbb80b74f"></a>
id0_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#a1c53ffb4e7e847387154745fbb80b74f">id0</a></td></tr>
<tr class="memdesc:a1c53ffb4e7e847387154745fbb80b74f"><td class="mdescLeft">&#160;</td><td class="mdescRight">scaffold number <br/></td></tr>
<tr class="separator:a1c53ffb4e7e847387154745fbb80b74f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a84bb4e4c9b6167a1cb9323439741b463"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a84bb4e4c9b6167a1cb9323439741b463"></a>
id1_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#a84bb4e4c9b6167a1cb9323439741b463">id1</a></td></tr>
<tr class="memdesc:a84bb4e4c9b6167a1cb9323439741b463"><td class="mdescLeft">&#160;</td><td class="mdescRight">location on scaffold <br/></td></tr>
<tr class="separator:a84bb4e4c9b6167a1cb9323439741b463"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-static-attribs"></a>
Static Public Attributes</h2></td></tr>
<tr class="memitem:afb3de39a8694b7c7d71fb78772a0e4ca"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="afb3de39a8694b7c7d71fb78772a0e4ca"></a>
static const std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#afb3de39a8694b7c7d71fb78772a0e4ca">table_name</a> =&quot;GENOTYPES&quot;</td></tr>
<tr class="memdesc:afb3de39a8694b7c7d71fb78772a0e4ca"><td class="mdescLeft">&#160;</td><td class="mdescRight">destination table in Db. <br/></td></tr>
<tr class="separator:afb3de39a8694b7c7d71fb78772a0e4ca"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0739b5ed675604551d5bc6700c9eca06"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0739b5ed675604551d5bc6700c9eca06"></a>
static const std::string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#a0739b5ed675604551d5bc6700c9eca06">file_name</a> =&quot;gcf&quot;</td></tr>
<tr class="memdesc:a0739b5ed675604551d5bc6700c9eca06"><td class="mdescLeft">&#160;</td><td class="mdescRight">defualt file extention. <br/></td></tr>
<tr class="separator:a0739b5ed675604551d5bc6700c9eca06"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pri-attribs"></a>
Private Attributes</h2></td></tr>
<tr class="memitem:a4ba690eaa16800aaa23fdf4cb973f278"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a4ba690eaa16800aaa23fdf4cb973f278"></a>
std::vector&lt; <a class="el" href="classgenotype.html">genotype</a> &gt;::iterator&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#a4ba690eaa16800aaa23fdf4cb973f278">igl_</a></td></tr>
<tr class="memdesc:a4ba690eaa16800aaa23fdf4cb973f278"><td class="mdescLeft">&#160;</td><td class="mdescRight">an iterator to allow us to iterate over the likelihoods. <br/></td></tr>
<tr class="separator:a4ba690eaa16800aaa23fdf4cb973f278"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a197ed37ae902b381c5d7cd3abad5e090"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a197ed37ae902b381c5d7cd3abad5e090"></a>
std::vector&lt; std::string &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#a197ed37ae902b381c5d7cd3abad5e090">sample_names_</a></td></tr>
<tr class="memdesc:a197ed37ae902b381c5d7cd3abad5e090"><td class="mdescLeft">&#160;</td><td class="mdescRight">a vector of sample names. <br/></td></tr>
<tr class="separator:a197ed37ae902b381c5d7cd3abad5e090"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8b05ae0bbd9945102669dda22a8ecb5c"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a8b05ae0bbd9945102669dda22a8ecb5c"></a>
bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classpopulation__genotypes.html#a8b05ae0bbd9945102669dda22a8ecb5c">frozen</a></td></tr>
<tr class="memdesc:a8b05ae0bbd9945102669dda22a8ecb5c"><td class="mdescLeft">&#160;</td><td class="mdescRight">a flag to indicate that no more samples will be added <br/></td></tr>
<tr class="separator:a8b05ae0bbd9945102669dda22a8ecb5c"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="a9fe492e927191dbf60f13e4965494a85"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t population_genotypes::size </td>
          <td>(</td>
          <td class="paramtype">void&#160;</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns the number of samples. </p>
<p>return size of <a class="el" href="classpopulation__genotypes.html">population_genotypes</a> if <a class="el" href="classpopulation__genotypes.html">population_genotypes</a> is set, 0 otherwise </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>mapgd_0.4/<a class="el" href="genotype_8h_source.html">genotype.h</a></li>
<li>mapgd_0.4/genotype.cc</li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Fri Feb 5 2016 10:05:47 for mapgd by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.6
</small></address>
</body>
</html>
