MapGD version 1.

<h1> To download this program click the "Download ZIP" button to your right -> </h1>

Copyright (C) Michael Lynch, see notice at end of README. At this time some of this document has been typed w/o Michael Lynch's review or approval. All typo's and inanities should be assigned to Matthew Ackerman until Michael Lynch has had time to review changes to this document. 

<h3> Introduction </h3>

This program uses a maximum-likelihood (ML) procedure to estimate the allele-frequencies from the numbers of the four nucleotides (quartets) observed at individual genomic sites. For each site, the major and minor nucleotides are identified, their frequencies are estimated by ML, and the polymorphism is tested for statistical significance.

The designated major nucleotide is simply the one with the highest rank, and the minor nucleotide is the one with the second highest rank. If the top three ranks are all equal, the site is treated as unresolvable, with both major and minor nucleotides designated in the output by a *.

If the second and third ranks are equal but lower than the major-nucleotide count, the site is treated as monomorphic, with the minor nucleotide again designated by a *.

<h3> Quick Start </h3>

<h5> Before you begin </h5>

This program is inteded to uses ".pro" files generated by the programs ace2pro, bow2pro, or sam2pro, which are all avalible ffrom http://guanine.evolbio.mpg.de/mlRho/

You may also find the programs samtools and bwa useful for aligning and analyzing your data. 

To download samtools please visit http://www.htslib.org/

To download bwa please visit http://bio-bwa.sourceforge.net/

Unfortunately these programs will not work in windows at this time (although CygWin may allow you to run them). Windows user may still use mapgd to analyze .pro files.  
	
<h5> For windows users </h5>

After downloading the program extract the file MapGD-master.zip to a location of your choice.
The program can be used by simply clicking on the "RunMeWin.bat" file in "scripts" subdirectory, which will analyze the "test.pro" file located in the test directory. 

mapgd can also be run from the command prompt, which can be accessed pressing the Windows logo key and r key simultaneously then typing "cmd" into menu which appears.

<h5> Linux or Mac users </h5>

Navigate to the directory that contains MapDG-master.zip, for example by typing “cd /home/LynchLab/Downloads/” 

Then type:

	unzip MapGD.zip
	cd MapGD-master
	Make

Scripts for both Linux and Mac users are present in the "scripts" subdirectory, or the program can be run by typing "mapdg -i FILENAME" where FILENAME is the a profile generated by one of the programs at the http://guanine.evolbio.mpg.de/mlRho

<h3> Details </h3>

<h5> Input file </h5>

The input file is a plain text file consists of six tab delimited columns, one for each site: the first two entries are arbitrary identifiers (e.g., scaffold number, and site), and the final four are integer values for the number of times an A, C, G, and T was observed at the site. 

If you wish to include additional columns for data identification, the following line must be edited in the file Estimator.cpp:

if (c!='>'){if (fscanf(instream,"%s\t%i\t%i\t%i\t%i", id2, &n[1], &n[2], &n[3], &n[4])==EOF) break;}

The program will try to open “datain.txt” if no file is specified. For example: “mapgd” will attempt to open datain.txt and print the analysis to the file "dataout.txt". If you desire to analyze a different file you can type “mapdg -i" and then the filename. 

<h5> Output file </h5>

Columns 1 and 2 are site identifiers; 3 and 4 designate major and minor nucleotides; 5,6 are the major- and minor-nucleotide frequencies; 7 is the estimated error rate; 8 is the total coverage at the site; 9 is the likelihood-ratio test statistic for polymorphism. Output columns are tab delimited.

Under the assumption of a chi-square distribution for the test statistic with one degree of freedom, significance at the 0.05, 0.01, 0.001 levels requires that the likelihood-ratio test statistic exceed 3.841, 6.635, and 10.827, respectively. 

In principle, the 95% support interval can be obtained by determining the changes in the estimate of the minor allele frequency in both directions required to reduce the log likelihood by the appropriate chi-square value (e.g., 3.841) although this is not currently implemented. 

By default the program prints information to the file "dataout.txt" and this file will appear in the same location as the program. If an alternative file name is desired simply type if "mapgd -o FILENAME" where FILENAME is the name of your output file.

<h3> Reference </h3>

Lynch, M., D. Bost, S. Wilson, T. Maruki, and S. Harrison. 2014. Population-genetic inference from pooled-sequencing data. Genome Biol. Evol. 6: 1210-1218.

<h3> Copyright Notice </h3>

This program is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

For a copy of the GNU General Public License write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
